homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,338 @@
+
+    for component in LEAF_COMPONENTS:
+        if (component != 'device_tracker' or
+                config[DOMAIN][CONF_NCONNECT] is True):
+            load_platform(hass, component, DOMAIN, {}, config)
+
+    return True
+
+
+class LeafDataStore:
+
+    def __init__(self, leaf, hass, config):
+        self.leaf = leaf
+        self.config = config
+        # self.nissan_connect = config[DOMAIN][CONF_NCONNECT]
+        self.nissan_connect = False  # Disabled until tested and implemented
+        self.force_miles = config[DOMAIN][CONF_FORCE_MILES]
+        self.hass = hass
+        self.data = {}
+        self.data[DATA_CLIMATE] = False
+        self.data[DATA_BATTERY] = 0
+        self.data[DATA_CHARGING] = False
+        self.data[DATA_LOCATION] = False
+        self.data[DATA_RANGE_AC] = 0
+        self.data[DATA_RANGE_AC_OFF] = 0
+        self.data[DATA_PLUGGED_IN] = False
+        self.last_check = None
+        track_time_interval(
+            hass,
+            self.refresh_leaf_if_necessary,
+            timedelta(seconds=CHECK_INTERVAL))
+
+    def refresh_leaf_if_necessary(self, event_time):
+        result = False
+        now = datetime.today()
+
+        base_interval = timedelta(minutes=self.config[DOMAIN][CONF_INTERVAL])
+        climate_interval = timedelta(
+            minutes=self.config[DOMAIN][CONF_CLIMATE_INTERVAL])
+        charging_interval = timedelta(
+            minutes=self.config[DOMAIN][CONF_CHARGING_INTERVAL])
+
+        if self.last_check is None:
+            _LOGGER.debug("Firing Refresh on %s"
+                          " as there has not been one yet.",
+                          self.leaf.nickname)
+            result = True
+        elif self.last_check + base_interval < now:
+            _LOGGER.debug("Firing Refresh on %s"
+                          " as the interval has passed.",
+                          self.leaf.nickname)
+            result = True
+        elif (self.data[DATA_CHARGING] is True and
+              self.last_check + charging_interval < now):
+            _LOGGER.debug("Firing Refresh on %s "
+                          " as it's charging and the charging interval"
+                          " has passed.",
+                          self.leaf.nickname)
+            result = True
+        elif (self.data[DATA_CLIMATE] is True and
+              self.last_check + climate_interval < now):
+            _LOGGER.debug("Firing Refresh on %s"
+                          " as climate control is on and"
+                          " the interval has passed.",
+                          self.leaf.nickname)
+            result = True
+
+        if result is True:
+            self.refresh_data()
+
+    def refresh_data(self):
+        _LOGGER.debug("Updating Nissan Leaf Data")
+
+        self.last_check = datetime.today()
+
+        battery_response = self.get_battery()
+        _LOGGER.debug("Got battery data for Leaf")
+
+        if battery_response.answer['status'] == 200:

```

Comment:
Why don't you store the `battery_response` object in the data store? That way you don't have to juggle around with all the DATA_ constants.

Reply:
How do you mean? I have lots of different responses from the car which is why I've got the constants.

Reply:
You can just store the latest response of every type

Reply:
OH, I see what you mean. Yeah that does make sense